--- a/klippy/chelper/__init__.py
+++ b/klippy/chelper/__init__.py
@@ -94,7 +94,7 @@ defs_trapq = """
         , double accel_t, double cruise_t, double decel_t
         , double start_pos_x, double start_pos_y, double start_pos_z
         , double axes_r_x, double axes_r_y, double axes_r_z
-        , double start_v, double cruise_v, double accel);
+        , double start_v, double cruise_v, double accel, uint32_t line);
     void trapq_finalize_moves(struct trapq *tq, double print_time
         , double clear_history_time);
     void trapq_set_position(struct trapq *tq, double print_time
@@ -170,6 +170,7 @@ defs_serialqueue = """
         uint8_t msg[MESSAGE_MAX];
         int len;
         double sent_time, receive_time;
+        uint64_t min_clock, req_clock;
         uint64_t notify_id;
     };
 
