--- a/klippy/configfile.py
+++ b/klippy/configfile.py
@@ -395,6 +395,7 @@ class PrinterConfig:
         if not self.autosave.fileconfig.sections():
             return
         gcode = self.printer.lookup_object('gcode')
+        is_restart = gcmd.get_int('RESTART', 1)
         # Create string containing autosave data
         autosave_data = self._build_config_string(self.autosave)
         lines = [('#*# ' + l).strip()
@@ -436,4 +437,5 @@ class PrinterConfig:
             logging.exception(msg)
             raise gcode.error(msg)
         # Request a restart
-        gcode.request_restart('restart')
+        if is_restart :
+            gcode.request_restart('restart')
