--- a/klippy/kinematics/cartesian.py
+++ b/klippy/kinematics/cartesian.py
@@ -61,6 +61,9 @@ class CartKinematics:
     def note_z_not_homed(self):
         # Helper for Safe Z Home
         self.limits[2] = (1.0, -1.0)
+    def note_xy_not_homed(self):
+        self.limits[0] = (1.0, -1.0)
+        self.limits[1] = (1.0, -1.0)
     def _home_axis(self, homing_state, axis, rail):
         # Determine movement
         position_min, position_max = rail.get_range()
--- a/klippy/kinematics/corexy.py
+++ b/klippy/kinematics/corexy.py
@@ -46,6 +46,9 @@ class CoreXYKinematics:
     def note_z_not_homed(self):
         # Helper for Safe Z Home
         self.limits[2] = (1.0, -1.0)
+    def note_xy_not_homed(self):
+        self.limits[0] = (1.0, -1.0)
+        self.limits[1] = (1.0, -1.0)
     def home(self, homing_state):
         # Each axis is homed independently and in order
         for axis in homing_state.get_axes():
--- a/klippy/kinematics/corexz.py
+++ b/klippy/kinematics/corexz.py
@@ -46,6 +46,9 @@ class CoreXZKinematics:
     def note_z_not_homed(self):
         # Helper for Safe Z Home
         self.limits[2] = (1.0, -1.0)
+    def note_xy_not_homed(self):
+        self.limits[0] = (1.0, -1.0)
+        self.limits[1] = (1.0, -1.0)
     def home(self, homing_state):
         # Each axis is homed independently and in order
         for axis in homing_state.get_axes():
--- a/klippy/kinematics/hybrid_corexy.py
+++ b/klippy/kinematics/hybrid_corexy.py
@@ -77,6 +77,9 @@ class HybridCoreXYKinematics:
     def note_z_not_homed(self):
         # Helper for Safe Z Home
         self.limits[2] = (1.0, -1.0)
+    def note_xy_not_homed(self):
+        self.limits[0] = (1.0, -1.0)
+        self.limits[1] = (1.0, -1.0)
     def _home_axis(self, homing_state, axis, rail):
         position_min, position_max = rail.get_range()
         hi = rail.get_homing_info()
--- a/klippy/kinematics/hybrid_corexz.py
+++ b/klippy/kinematics/hybrid_corexz.py
@@ -77,6 +77,9 @@ class HybridCoreXZKinematics:
     def note_z_not_homed(self):
         # Helper for Safe Z Home
         self.limits[2] = (1.0, -1.0)
+    def note_xy_not_homed(self):
+        self.limits[0] = (1.0, -1.0)
+        self.limits[1] = (1.0, -1.0)
     def _home_axis(self, homing_state, axis, rail):
         position_min, position_max = rail.get_range()
         hi = rail.get_homing_info()
--- a/klippy/kinematics/polar.py
+++ b/klippy/kinematics/polar.py
@@ -54,6 +54,9 @@ class PolarKinematics:
     def note_z_not_homed(self):
         # Helper for Safe Z Home
         self.limit_z = (1.0, -1.0)
+    def note_xy_not_homed(self):
+        self.limits[0] = (1.0, -1.0)
+        self.limits[1] = (1.0, -1.0)
     def _home_axis(self, homing_state, axis, rail):
         # Determine movement
         position_min, position_max = rail.get_range()
